### PR TITLE
Incremental testing 02bc 03ab

### PR DESCRIPTION
**Pull request by crtrott**

### CODE REVIEW

**Comment by crtrott** [comment]
This is replacing #2619 #2618 #2538

File: cmake/kokkos_test_cxx_std.cmake [hunk 1]
```diff
@@ -14,12 +14,12 @@ FUNCTION(kokkos_set_cxx_standard_feature standard)
...
   ELSEIF(CMAKE_CXX_EXTENSIONS)
     IF(KOKKOS_DONT_ALLOW_EXTENSIONS)
-      MESSAGE(FATAL_ERROR "The chosen configuration does not support CXX extensions flags: ${KOKKOS_DONT_ALLOW_EXTENSIONS}. Must set CMAKE_CXX_EXTENSIONS=OFF to continue")
+      MESSAGE(FATAL_ERROR "The chosen configuration does not support CXX extensions flags: ${KOKKOS_DONT_ALLOW_EXTENSIONS}. Must set CMAKE_CXX_EXTENSIONS=OFF to continue") 
     ELSE()
       GLOBAL_SET(KOKKOS_USE_CXX_EXTENSIONS ON)
     ENDIF()
   ELSE()
-    #For trilinos, we need to make sure downstream projects
+    #For trilinos, we need to make sure downstream projects
```
Comment (dalg24) [inline cmake/kokkos_test_cxx_std.cmake]: Do not add trailing whitespaces

File: cmake/kokkos_test_cxx_std.cmake [hunk 2]
```diff
@@ -35,18 +35,9 @@ FUNCTION(kokkos_set_cxx_standard_feature standard)
...
-                                    MESSAGE(FATAL_ERROR "Compiler ${KOKKOS_CXX_COMPILER_ID} should support ${FEATURE_NAME}, but CMake reports feature not supported")
-                                        ENDIF()
-                                          ENDIF()
-
-  #  IF(NOT ${FEATURE_NAME} IN_LIST CMAKE_CXX_COMPILE_FEATURES)
-  #    MESSAGE(FATAL_ERROR "Compiler ${KOKKOS_CXX_COMPILER_ID} should support ${FEATURE_NAME}, but CMake reports feature not supported")
-  #  ENDIF()
+  IF(NOT ${FEATURE_NAME} IN_LIST CMAKE_CXX_COMPILE_FEATURES)
+    MESSAGE(FATAL_ERROR "Compiler ${KOKKOS_CXX_COMPILER_ID} should support ${FEATURE_NAME}, but CMake reports feature not supported")
+  ENDIF()
```
Comment (dalg24) [inline cmake/kokkos_test_cxx_std.cmake]: You are setting yourself up for conflicts when we will merge into develop.  You should not be editing that file at all.

File: core/unit_test/incremental/Test02c_MemorySpace.hpp [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+// DeepCopy unit tests
+// We allocate data in the host. Copy it to the device and copy back the data
+// from device back to host We check if the original and copied data are the
+// same to evaluate the deep copies.
+
+template <class ExecSpace>
+struct TestIncrMemorySpace_deepcopy {
+  using dataType         = double;
+  const int num_elements = 10;
+  const double value     = 0.5;
```
Comment (dalg24) [inline core/unit_test/incremental/Test02c_MemorySpace.hpp]: If you want the type alias at all then fix the style `dataType -> DataType` and use it when you declare `value`.

File: core/unit_test/incremental/Test02c_MemorySpace.hpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+  // Memory Space for Device and Host data
+  typedef typename ExecSpace::memory_space memSpaceD;
+  typedef Kokkos::HostSpace memSpaceH;
```
Comment (dalg24) [inline core/unit_test/incremental/Test02c_MemorySpace.hpp]: This is outdated. I feel like I have reviewed this code before and made the same comments.

File: cmake/kokkos_test_cxx_std.cmake (hunk 2 above)
Comment (crtrott) [reply cmake/kokkos_test_cxx_std.cmake]: I reverted a change which Jeremiah said shouldn't have happened in the first place. This is not a conflict.

File: cmake/kokkos_test_cxx_std.cmake (hunk 1 above)
Comment (crtrott) [reply cmake/kokkos_test_cxx_std.cmake]: This is technically not adding a whitespace, this commit reverts a change which should not have happened. The whitespace is probably present in develop.

Comment (crtrott) [reply cmake/kokkos_test_cxx_std.cmake]: Yeah just checked the whitespace is present in develop. This file is restored to the exact version in develop.

File: core/unit_test/incremental/Test02c_MemorySpace.hpp (hunk 4 above)
Comment (crtrott) [reply core/unit_test/incremental/Test02c_MemorySpace.hpp]: replaced with using statement and naming convention more in line with rest of Kokkos

File: core/unit_test/incremental/Test02c_MemorySpace.hpp (hunk 3 above)
Comment (crtrott) [reply core/unit_test/incremental/Test02c_MemorySpace.hpp]: fixed naming convention and value declaration

**Comment by crtrott** [comment]
force push amended commit with formatting fix

**Comment by rgayatri23** [comment]
Retest this please.

File: core/unit_test/incremental/Test02c_MemorySpace.hpp [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+const value_type value = 0.5;
+
+template <class ExecSpace>
+struct TestIncrMemorySpace_deepcopy {
+  // Memory space type for Device and Host data
+  using d_memspace_type = typename ExecSpace::memory_space;
+  using h_memspace_type = Kokkos::HostSpace;
+
+  // Routine to check correcness.
+  // Checks whether the original host array and the array copied/recieved from
```
Comment (masterleinad) [inline core/unit_test/incremental/Test02c_MemorySpace.hpp]:
```suggestion
  // Checks whether the original host array and the array copied/received from
```

File: core/unit_test/incremental/Test02c_MemorySpace.hpp [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    auto *hostData_recv = allocate_mem<h_memspace_type>(num_elements);
+    ASSERT_NE(hostData_recv, nullptr);
+
+    // Initialize Host data buffers with different values.
+    for (int i = 0; i < num_elements; ++i) {
+      hostData[i]      = (i + 1) * value;
+      hostData_recv[i] = num_elements - i;
+    }
+
+    // Copy first from Host_send to Device.
```
Comment (masterleinad) [inline core/unit_test/incremental/Test02c_MemorySpace.hpp]:
```suggestion
    // Copy first from Host to Device.
```

File: core/unit_test/incremental/Test02c_MemorySpace.hpp [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    Kokkos::Impl::DeepCopy<h_memspace_type, d_memspace_type>(
+        deviceData, hostData, num_elements * sizeof(value_type));
+
+    // Copy from device to the recieve buffer of the Host.
```
Comment (masterleinad) [inline core/unit_test/incremental/Test02c_MemorySpace.hpp]:
```suggestion
    // Copy from Device to the receive buffer of the Host.
```

File: core/unit_test/incremental/Test02c_MemorySpace.hpp [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+*/
+
+#include <Kokkos_Core.hpp>
+#include <gtest/gtest.h>
+
+/// @Kokkos_Feature_Level_Required:2
+
+// Deep copy unit tests
+// We allocate and initialize an array on the Host.
+// Copy the array to the Device and copy back the same from Device To Host in a
```
Comment (masterleinad) [inline core/unit_test/incremental/Test02c_MemorySpace.hpp]:
```suggestion
// Copy the array to the Device and copy back the same from Device to Host in a
```

File: core/unit_test/incremental/Test03a_ParallelFor.hpp [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+// ************************************************************************
+//@HEADER
+*/
+
+#include <Kokkos_Core.hpp>
+#include <gtest/gtest.h>
+
+/// @Kokkos_Feature_Level_Required:3
+// parallel-for unit test.
+// In this test different elements of an array are updated by different threads.
```
Comment (masterleinad) [inline core/unit_test/incremental/Test03a_ParallelFor.hpp]:
```suggestion
// In this test, different elements of an array are updated by different threads.
```

File: core/unit_test/incremental/Test03a_ParallelFor.hpp [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+
+// A macro representing the values assigned to the elements of the array.
+// This macro is used while updating the array and while checking the
+// correctness.
+#define updateValue (i + 1) * value
```
Comment (masterleinad) [inline core/unit_test/incremental/Test03a_ParallelFor.hpp]: Why do we need this at all? I would avoid macros as much as possible. If necessary, you can make it a lambda.

File: core/unit_test/incremental/Test03a_ParallelFor.hpp [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+template <class ExecSpace>
+struct TestParallel_For {
+  // Memory space type for Device and Host data
+  using d_memspace_type = typename ExecSpace::memory_space;
+  using h_memspace_type = Kokkos::HostSpace;
+
+  value_type *deviceData, *hostData;
+
+  // Check if the array values are updated correctly
```
Comment (masterleinad) [inline core/unit_test/incremental/Test03a_ParallelFor.hpp]:
```suggestion
  // Check if the array values are updated correctly.
```

File: core/unit_test/incremental/Test03a_ParallelFor.hpp [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+          << "Values in index " << i << " are incorrect";
+    }
+  }
+
+  // Routine to allocate memory in a specific memory space.
+  template <class MemSpace>
+  value_type *allocate_mem(int N) {
+    return (static_cast<value_type *>(
+        Kokkos::kokkos_malloc<MemSpace>("deviceData", N * sizeof(value_type))));
+  }
```
Comment (masterleinad) [inline core/unit_test/incremental/Test03a_ParallelFor.hpp]: We could consider moving these into `Kokkos` (but certainly not in this pull request).

File: core/unit_test/incremental/Test03b_Reduction.hpp [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <Kokkos_Core.hpp>
+#include <gtest/gtest.h>
+
+/// @Kokkos_Feature_Level_Required:3
+// Unit test for reduction of native data type
```
Comment (masterleinad) [inline core/unit_test/incremental/Test03b_Reduction.hpp]:
```suggestion
// Unit test for reduction of native data type.
```

File: core/unit_test/incremental/Test03b_Reduction.hpp [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+const int num_elements = 10;
+
+struct ReduceFunctor {
+  value_type *_data;
+
+  ReduceFunctor(value_type *data) : _data(data) {}
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const int i, double &UpdateSum) const {
+    _data[i] = (i + 1) * value;
```
Comment (masterleinad) [inline core/unit_test/incremental/Test03b_Reduction.hpp]: I like that better than the macro in the other test.

File: core/unit_test/incremental/Test03a_ParallelFor.hpp (hunk 10 above)
Comment (rgayatri23) [reply core/unit_test/incremental/Test03a_ParallelFor.hpp]: I had the macro for this here, because if we later felt like the value updated needs to be modified, we can do it at one place rather than two places (while updating and during the correctness check). I don't have any strong feelings about it one way or the other, we can change it if that is not a good idea.

**Comment by rgayatri23** [comment]
> Really just nitpicks for consistency. I am fine with this pull request if you replace the macro.

Done. Also fixed the spelling mistakes

File: core/unit_test/incremental/Test03a_ParallelFor.hpp (hunk 12 above)
Comment (crtrott) [reply core/unit_test/incremental/Test03a_ParallelFor.hpp]: What do you mean? What exactly do you want which is different from kokkos_malloc?

Comment (masterleinad) [reply core/unit_test/incremental/Test03a_ParallelFor.hpp]: Something like a version templated on the type to be allocated, such that we can avoid casting the resulting pointer and can pass the number of objects instead of the size in bytes.

File: cmake/kokkos_test_cxx_std.cmake [hunk 15]
```diff
@@ -35,9 +35,18 @@ FUNCTION(kokkos_set_cxx_standard_feature standard)
...
+                  MESSAGE(STATUS "nvcc_wrapper does not support TARGET_COMPILE_FEATURES")
+                        GLOBAL_SET(KOKKOS_CXX_STANDARD_FEATURE "")
+                            ELSE()
+                                    MESSAGE(FATAL_ERROR "Compiler ${KOKKOS_CXX_COMPILER_ID} should support ${FEATURE_NAME}, but CMake reports feature not supported")
+                                        ENDIF()
+                                          ENDIF()
+
+  #  IF(NOT ${FEATURE_NAME} IN_LIST CMAKE_CXX_COMPILE_FEATURES)
+  #    MESSAGE(FATAL_ERROR "Compiler ${KOKKOS_CXX_COMPILER_ID} should support ${FEATURE_NAME}, but CMake reports feature not supported")
+  #  ENDIF()
```
Comment (masterleinad) [inline cmake/kokkos_test_cxx_std.cmake]: This looks different from `develop`. Is this change intended?

File: core/unit_test/incremental/Test02c_MemorySpace.hpp [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+      ASSERT_EQ(hostData[i], hostData_recv[i])
+          << "hostData and hostData_recv differ at index" << i;
+    }
+  }
+
+  // Routine to allocate memory in a specific memory space.
+  template <class MemSpace>
+  value_type *allocate_mem(int N) {
+    return (static_cast<value_type *>(
+        Kokkos::kokkos_malloc<MemSpace>("deviceData", N * sizeof(value_type))));
```
Comment (masterleinad) [inline core/unit_test/incremental/Test02c_MemorySpace.hpp]: Technically, you are not always allocating device data using this function.

File: core/unit_test/incremental/Test02c_MemorySpace.hpp [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+    Kokkos::Impl::DeepCopy<d_memspace_type, h_memspace_type>(
+        hostData_recv, deviceData, num_elements * sizeof(value_type));
+
+    // Check if all data has been copied correctly back to the Host.
+    correctness_check(hostData, hostData_recv);
+
+    // Free the allocated memory.
+    free_mem<d_memspace_type>(deviceData);
+    free_mem<h_memspace_type>(hostData);
+    free_mem<h_memspace_type>(hostData_recv);
```
Comment (masterleinad) [inline core/unit_test/incremental/Test02c_MemorySpace.hpp]: Is this better than calling `Kokkos::kokkos_free<MemSpace>(data);` directly?

File: core/unit_test/incremental/Test03a_ParallelFor.hpp [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+  void check_correctness_and_cleanup() {
+    // Copy the data back to Host memory space
+    Kokkos::Impl::DeepCopy<d_memspace_type, h_memspace_type>(
+        hostData, deviceData, num_elements * sizeof(value_type));
+
+    // Check if all data has been update correctly
+    correctness_check(hostData);
+
+    // free the allocated memory
+    free_mem<d_memspace_type>(deviceData);
```
Comment (masterleinad) [inline core/unit_test/incremental/Test03a_ParallelFor.hpp]: See above.

File: core/unit_test/incremental/Test03a_ParallelFor.hpp [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+    hostData = allocate_mem<h_memspace_type>(num_elements);
+    ASSERT_NE(hostData, nullptr);
+  }
+
+  void check_correctness_and_cleanup() {
+    // Copy the data back to Host memory space
+    Kokkos::Impl::DeepCopy<d_memspace_type, h_memspace_type>(
+        hostData, deviceData, num_elements * sizeof(value_type));
+
+    // Check if all data has been update correctly
```
Comment (masterleinad) [inline core/unit_test/incremental/Test03a_ParallelFor.hpp]:
```suggestion
    // Check if all data has been updated correctly
```

File: cmake/kokkos_test_cxx_std.cmake (hunk 15 above)
Comment (rgayatri23) [reply cmake/kokkos_test_cxx_std.cmake]: I think this was updated in develop after we created the incremental-testing branch.

File: core/unit_test/incremental/Test02c_MemorySpace.hpp (hunk 17 above)
Comment (rgayatri23) [reply core/unit_test/incremental/Test02c_MemorySpace.hpp]: I had a different routine for free (even though it is just one line) here so that it is consistent with having a routine for kokkos_malloc too.

File: cmake/kokkos_test_cxx_std.cmake (hunk 15 above)
Comment (masterleinad) [reply cmake/kokkos_test_cxx_std.cmake]: It seems that this change is also not present looking at all the commits in common, but it was in one of the recent comments. Again, cleaning up the history makes reviewing so much easier.

File: core/unit_test/incremental/Test02c_MemorySpace.hpp (hunk 16 above)
Comment (rgayatri23) [reply core/unit_test/incremental/Test02c_MemorySpace.hpp]: Do you mean there is a data allocation that is not happening via "allocate_mem" routine?

Comment (masterleinad) [reply core/unit_test/incremental/Test02c_MemorySpace.hpp]: No, I am saying that the label "deviceData" might be a bit misleading since you also use this function to allocate memory in host space.

Comment (rgayatri23) [reply core/unit_test/incremental/Test02c_MemorySpace.hpp]: Ah yes. Fixed it.

**Comment by masterleinad** [comment]
Why are there still the other 13 commits? Also, there seems to be a merge conflict now.

File: cmake/kokkos_test_cxx_std.cmake (hunk 15 above)
Comment (crtrott) [reply cmake/kokkos_test_cxx_std.cmake]: where is this coming from? 
Why was this pushed here again.

File: core/unit_test/incremental/Test03b_Reduction.hpp [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+  }
+
+  void sum_reduction() {
+    // Allocates memory for num_elements number of value_type elements in the
+    // host and device memory spaces.
+    init();
+
+    // Creates a range policy that uses static schedule for Cuda and dynamic
+    // schedule for other memory spaces.
+#if defined(KOKKOS_ENABLE_CUDA)
```
Comment (crtrott) [inline core/unit_test/incremental/Test03b_Reduction.hpp]: Where is this coming from? I thought I had fixed this?

Comment (crtrott) [reply core/unit_test/incremental/Test03b_Reduction.hpp]: Was a different file I guess. Anyway, there is no reason to do this here. `Kokkos::Dynamic` is totally fine to request from CUDA too.

Comment (rgayatri23) [reply core/unit_test/incremental/Test03b_Reduction.hpp]: Modified the test to use Kokkos::Dynamic always.
